tidy3d/plugins/microwave/custom_path_integrals.py

```diff
@@ -0,0 +1,163 @@
+    ) -> IntegralResultTypes:
+        """Computes the path integral defined by ``vertices`` given the input ``em_field``.
+        The ``FieldParameter`` determines whether the electric or magnetic field is integrated."""
+        if not isinstance(em_field, (FieldData, FieldTimeData, ModeSolverData)):
+            raise DataError("'em_field' type not supported.")
+
+        dim1 = "xyz"[self.remaining_axes[0]]
+        dim2 = "xyz"[self.remaining_axes[1]]
+        dim3 = "xyz"[self.main_axis]
+
+        h_field_name = f"{field}{dim1}"
+        v_field_name = f"{field}{dim2}"
+        # Validate that fields are present
+        if h_field_name not in em_field.field_components:
+            raise DataError(f"'field_name' '{h_field_name}' not found.")
+        if v_field_name not in em_field.field_components:
+            raise DataError(f"'field_name' '{v_field_name}' not found.")
+
+        # Select fields lying on the plane
+        plane_indexer = {dim3: self.position}
+        field1 = em_field.field_components[h_field_name].sel(plane_indexer, method="nearest")
+        field2 = em_field.field_components[v_field_name].sel(plane_indexer, method="nearest")
+
+        # Dim 's' represents the parameterization of the line
+        # 't' is likely used for time
+        x_path = xr.DataArray(self.vertices[:, 0], dims="s")
+        y_path = xr.DataArray(self.vertices[:, 1], dims="s")
+        field1_interp = field1.interp(x=x_path, y=y_path, method="linear")
+        field2_interp = field2.interp(x=x_path, y=y_path, method="linear")
+
+        # Determine the differential length elements along the path
+        dl_x = self._compute_dl_component(x_path, self.is_closed_contour)
+        dl_y = self._compute_dl_component(y_path, self.is_closed_contour)
+        dl_x = xr.DataArray(dl_x, dims="s")
+        dl_y = xr.DataArray(dl_y, dims="s")
+
+        # Compute the dot product between differential length element and vector field
+        integrand = field1_interp * dl_x + field2_interp * dl_y
+        # Integrate along the path
+        result = integrand.integrate(coord="s")
+        result = result.reset_coords(drop=True)
+
+        if isinstance(em_field, FieldData):
+            return FreqDataArray(data=result.data, coords=result.coords)
+        elif isinstance(em_field, FieldTimeData):
+            return TimeDataArray(data=result.data, coords=result.coords)
+        else:
+            assert isinstance(em_field, ModeSolverData)
+            return FreqModeDataArray(data=result.data, coords=result.coords)
+
+    @staticmethod
+    def _compute_dl_component(coord_array: xr.DataArray, closed_contour=False) -> np.array:
+        """Computes the differential length element along the integration path."""
+        dl = np.gradient(coord_array)
+        if closed_contour:
+            end_points = (coord_array[1] - coord_array[-2]).values / 2
+            dl[0] = end_points
+            dl[-1] = end_points
```

Comment:
explain these 3 Ines? a little confused here.

Reply:
for example, why is `coord_array[0]` and `coord_array[-1]` ignored?

Reply:
I'll update the code with comments, but if you are interested it is because I would like to use a central difference scheme to approximate the derivative at the first and last point when the contour is closed. (The first and last point have identical positions for my definition of a closed contour). So I need the second point and the penultimate point to compute the values for the zeroth point(which is the same as the last point).

Reply:
ok I see. I guess it might be worth noting that np.gradient probably uses 3 points in it's calculation (https://numpy.org/doc/stable/reference/generated/numpy.gradient.html) so is it worth doing this for the end point too? It seems this approximation kind of ignores `coord_array[0]`, which could make it less accurate?

Reply:
Maybe a quick and easy fix could be something like
```py
if closed_contour:
    grad_end = np.gradient([coord_array[-2], coord_array[0],  coord_array[1]])
    dl[0] = dl[-1] = grad_end

```

Reply:
Well both methods give the same result (the last equation of the note that you linked is what I was emulating). However, I will concede that your fix is nicer looking :smile: !